keps/sig-api-machinery/3157-watch-list/README.md

```diff
@@ -0,0 +1,991 @@
+The bottom line is around O(5*the_response_from_etcd) of temporary memory consumption.
+Neither priority and fairness nor Golang garbage collection is able to protect the system from exhausting memory.
+
+A situation like that is dangerous twofold.
+First, as we saw it could slow down if not fully stop an API server that has received the requests.
+Secondly, a sudden and uncontrolled spike in memory consumption will likely put pressure on the node itself.
+This might lead to thrashing, starving, and finally losing other processes running on the same node, including kubelet.
+Stopping kubelet has serious issues as it leads to workload disruption and a much bigger blast radius.
+Note that in that scenario even clusters in an HA setup are affected.
+
+Worse, in rare cases (see the [Appendix](#appendix) section for more) recovery of large clusters with therefore many kubelets and hence informers for pods, secrets, configmap can lead to a very expensive storm of LISTs.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+- protect kube-apiserver and its node against list-based OOM attacks
+- considerably reduce (temporary) memory footprint of LISTs, down from O(watchers\*page-size\*object-size\*5) to O(watchers\*constant), constant around 2 MB.
+>  Example:
+> 
+>  512 watches of 400mb data: 512*500*2MB*5=2.5TB ↘ 2 GB
+> 
+>  racing with Golang GC to free this temporary memory before being OOM'ed.
+
+- reduce etcd load by serving from watch cache
+- get a replacement for paginated lists from watch-cache, which is not feasible without major investment
+- enforce consistency in the sense of freshness of the returned list
+- be backward compatible with new client -> old server
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- get rid of list or list pagination
+- rewrite the list storage stack to allow streaming, but rather use the existing streaming infrastructure (watches).
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+In order to lower memory consumption while getting a list of data and make it more predictable, we propose to use consistent streaming from the watch-cache instead of paging from etcd.
+Initially, the proposed changes will be applied to informers as they are usually the heaviest users of LIST requests (see [Appendix](#appendix) section for more details on how informers operate today).
+The primary idea is to use standard WATCH request mechanics for getting a stream of individual objects, but to use it for LISTs.
```

Comment:
Are you actually changing LIST or are you changing informers (client) to use a watch and then tweaking watches?

Reply:
we are going to change informers to use a single WATCH request instead of a LIST followed by a WATCH request.

Reply:
If the primary problem is that apiserver can run out of memory, fixing clients is not going to solve that problem. (I haven't read this yet)